File: apps/web/lib/queryRowCounts.ts
import { runWarehouseSql } from "@runoff/core";
import type { BlueprintContent } from "@runoff/core";
import type { getDb } from "./db";

/**
 * Per-query row counts for a blueprint's baked section queries, bound to the
 * project's latest filed period. Shape: sectionKey → query name → count
 * (`null` when the query fails to compile or run). Sections without queries are
 * omitted. Used by the Builder to render read-only `<name> — <n> rows` chips.
 */
export function computeQueryRowCounts(
  db: ReturnType<typeof getDb>,
  projectId: string,
  content: BlueprintContent,
): Record<string, Record<string, number | null>> {
  const latest = (
    db.sqlite
      .prepare(
        "SELECT MAX(period) AS p FROM sources WHERE project_id = ? AND status='filed' AND period IS NOT NULL",
      )
      .get(projectId) as { p: string | null }
  ).p;

  const queryRowCounts: Record<string, Record<string, number | null>> = {};
  for (const s of content.sections) {
    if (!s.queries.length) continue;
    queryRowCounts[s.key] = {};
    for (const qy of s.queries) {
      try {
        const res = runWarehouseSql(projectId, `SELECT COUNT(*) FROM (${qy.sql})`, { period: latest });
        queryRowCounts[s.key][qy.name] = (res.rows[0]?.[0] as number) ?? null;
      } catch {
        queryRowCounts[s.key][qy.name] = null;
      }
    }
  }
  return queryRowCounts;
}

File: backend/runoff_api/services/query_row_counts.py
"""Port of apps/web/lib/queryRowCounts.ts — computeQueryRowCounts.

Per-query row counts for a blueprint's baked section queries, bound to the
project's latest filed period. Shape: sectionKey -> query name -> count (None
when the query fails to compile or run). Sections without queries are omitted.
"""

from runoff_api.core.db import RunoffDb
from runoff_api.core.warehouse import run_warehouse_sql


def compute_query_row_counts(
    db: RunoffDb, project_id: str, content: dict
) -> dict[str, dict[str, int | None]]:
    latest = db.execute(
        "SELECT MAX(period) AS p FROM sources WHERE project_id = ? AND status='filed' AND period IS NOT NULL",
        (project_id,),
    ).fetchone()["p"]

    query_row_counts: dict[str, dict[str, int | None]] = {}
    for s in content["sections"]:
        if not s["queries"]:
            continue
        query_row_counts[s["key"]] = {}
        for qy in s["queries"]:
            try:
                res = run_warehouse_sql(project_id, f"SELECT COUNT(*) FROM ({qy['sql']})", latest)
                rows = res["rows"]
                count = rows[0][0] if rows else None
                query_row_counts[s["key"]][qy["name"]] = count if count is not None else None
            except Exception:  # noqa: BLE001 — any compile/run error => null count
                query_row_counts[s["key"]][qy["name"]] = None
    return query_row_counts
